packages/nitro-protocol/contracts/examples/EmbeddedApplication.sol

```diff
@@ -0,0 +1,446 @@
+                    'incorrect move from B'
+                );
+            } else {
+                revert('move from None,A,B only');
+            }
+        }
+
+        // validate the supplied support proof
+        // extract the allocation
+        Outcome.AllocationItem[] memory Xallocation = validateSupportProofForX(
+            fromAppData,
+            toAppData
+        );
+
+        // ensure A,B part of the outcome of X has been absorbed into the outcome of J
+        require(
+            Xallocation[AIndex].amount == toAllocation[AIndex].amount &&
+                Xallocation[BIndex].amount == toAllocation[BIndex].amount &&
+                Xallocation[AIndex].destination == toAllocation[AIndex].destination &&
+                Xallocation[BIndex].destination == toAllocation[BIndex].destination,
+            'X / J outcome mismatch'
+        );
+        return true;
+    }
+
+    function validateSupportProofForX(AppData memory fromAppData, AppData memory toAppData)
+        internal
+        pure
+        returns (Outcome.AllocationItem[] memory Xallocation)
+    {
+        // The following checks follow the protocol-level validTransition function
+        // They are the members of FixedPart that do not affect the channelId
+        require(
+            fromAppData.supportProofForX.fixedPart.appDefinition ==
+                toAppData.supportProofForX.fixedPart.appDefinition,
+            'X.appDefinition changed'
+        );
+        require(
+            fromAppData.supportProofForX.fixedPart.challengeDuration ==
+                toAppData.supportProofForX.fixedPart.challengeDuration,
+            'X.challengeDuration changed'
+        ); // this is actually never used so could be 0
+
+        // A support proof requires 2 signatures
+        // But it may have either 1 or two states.
+        // If both signatures on the same state, it is supported
+        // Else one signature per state and we must check for a valid transition
+        require(
+            toAppData.supportProofForX.variableParts.length == 1 ||
+                toAppData.supportProofForX.variableParts.length == 2,
+            '1 or 2 states required'
+        );
+
+        bytes32 appPartHash = keccak256(
+            abi.encode(
+                toAppData.supportProofForX.fixedPart.challengeDuration,
+                toAppData.supportProofForX.fixedPart.appDefinition,
+                toAppData.supportProofForX.variableParts[0].appData
+            )
+        );
+
+        // hash the greatest state first (either the later of a pair, or the only state provided)
+
+        uint256 finalIndex = toAppData.supportProofForX.variableParts.length - 1;
+
+        bytes memory greaterStateOutcome = toAppData.supportProofForX.variableParts[finalIndex]
+            .outcome;
+
+        bytes32 greaterStateHash = keccak256(
+            abi.encode(
+                IForceMove.State(
+                    toAppData.supportProofForX.turnNumTo,
```

Comment:
`toAppData.supportProofForX` is referenced many times. Should we name this as a variable? Maybe something like `toSupportProof`?

Reply:
We could, but I'm a bit skeptical about the benefit. Is `toSupportProofForX.turnNumTo` plus an extra variable declaration really any better than `toAppData.supportProofForX.turnNumTo` ? If it is, would we stop there or also have `toSupportProofVariableParts[0]` and `toSupportProofFixedPart.participant[0]` or even `toSupportProofParticipants[0]` plus many more variables?

Personally I like seeing where each function argument comes from explicitly. I think naming new variables increases the chances of bugs.

Reply:
> Personally I like seeing where each function argument comes from explicitly. I think naming new variables increases the chances of bugs.

In this case, we want to validate many things about `toAppData.supportProofForX`. Assigning its value to a variable, _in theory_, prevents mistakes like
```
            (toAppData.supportProofForX.whoSignedWhat[0] == 0) &
            (fromAppData.supportProofForX.whoSignedWhat[1] == 0)
```
though I admit that this mistake might be easy to catch?

My suggestion is mostly motivated by improving the readability of all of the places where we are validating things about `toAppData.supportProofForX`. If we call it `nextSupportProof`, it makes it easier for a reader to "keep everything in ones head" -- instead of having to check:
- which state's app data does this come from?
- what are we checking about the app data?

you are indicating: "I'm going to focus on the `nextSupportProof`, which is `toAppData.supportProofForX`. I am going to check these properties of the next support proof."

I'm going to try to understand the properties verified below as is, and then come back to this suggestion!